### PR TITLE
Support views API

### PR DESCRIPTION
**Pull request by kanata2**
### Overview
This PR supports views API. This is useful for Modal, Home Tab.

ref: https://api.slack.com/reference/surfaces/views

---

##### Pull Request Guidelines

These are recommendations for pull requests.  
They are strictly guidelines to help manage expectations.

##### should this be an issue instead
- [ ] is it a convenience method? (no new functionality, streamlines some use case)
- [ ] exposes a previously private type, const, method, etc.
- [ ] is it application specific (caching, retry logic, rate limiting, etc)
- [ ] is it performance related.

##### API changes

Since API changes have to be maintained they undergo a more detailed review and are more likely to require changes.

- no tests, if you're adding to the API include at least a single test of the happy case.
- If you can accomplish your goal without changing the API, then do so.
- dependency changes. updates are okay. adding/removing need justification.

###### examples of API changes that do not meet guidelines:
- in library cache for users. caches are use case specific.
- Convenience methods for Sending Messages, update, post, ephemeral, etc. consider opening an issue instead.

### CODE REVIEW

File: views.go
Comment (kanata2) [inline views.go:26]: NOTE: The type of `state` is not clear (not used) so far.

File: slackevents/inner_events.go
Comment (kanata2) [inline slackevents/inner_events.go:34]: ref: https://api.slack.com/surfaces/tabs

**Comment by russellseymour** [comment]
I am writing a bot for Slack and want to use Modal dialog boxes. This PR is a great addition :+1:

Is there a timescale on when it might be merged?

**Comment by kanata2** [comment]
Thanks, @russellseymour :)

@james-lawrence Would you check this PR when you're available?

**Comment by tsein-stripe** [comment]
@kanata2 This is just what I was looking for! Thank you for taking the time to make this contribution, I can't wait til it's merged in 😄

**Comment by tsein-stripe** [comment]
@kanata2 one small request, could you include the two new [view interaction types](https://api.slack.com/reference/interaction-payloads/views), `view_submission` and `view_closed` here:
https://github.com/nlopes/slack/blob/master/interactions.go#L20-L27

You'll also want to add a view field, for CallbackID matching on submission/closed interactions here:
https://github.com/nlopes/slack/blob/master/interactions.go#L46

**Comment by kanata2** [comment]
@tsein-stripe
Thank you for your request! ~~I'll add such types later.~~
[updated] I am planning to handle for user's interactions against views in next PR.

File: views.go
Comment (ollieparsley) [inline views.go]: @kanata2 I've been trying this out and I came across an issue due to `hash` being optional https://api.slack.com/methods/views.publish . If you supply an empty string as the hash value (in order to not send the hash value) the API request still includes the empty string, this causes Slack to give the `hash_conflict` error if you are trying to publish a view that already has a hash defined. The fix is easy though! Omit the attribute from the API request payload if it's empty. This fix also applies to the `ViewUpdate` method too (separate struct).
```
Hash   string             `json:"hash,omitempty"`
```

Also, nice job on this!

Comment (kanata2) [reply views.go]: Thank you for your review! I missed it...
Handle optional values as `omitempty` in c14171d

File: views.go
Comment (tsein-stripe) [reply views.go:26]: The `state` is used in `view_submission` interaction payloads. It's a dictionary of dictionaries, the top level dictionary contains BlockID keys and the nested dictionary contains ActionID keys which map the type and value of the user input on a block element:
https://api.slack.com/reference/interaction-payloads/views#view_submission

I'm using the following in my fork to represent that:
```
type ElementState struct {
	Type  string `json:"type"`
	Value string `json:"value"`
}

type ViewState struct {
	Values map[string]map[string]ElementState `json:"values"`
}
```

Which supports the following JSON structure:
```
"state": {
    "values": {
        "multi-line": {
            "ml-value": {
                "type": "plain_text_input",
                "value": "This is my example inputted value"
            }
        }
    }
}
```

Comment (kanata2) [reply views.go:26]: Good catch! Thanks!
I think this usage of `state` is recently update and Slack updates its API day by day.
So I want to merge this PR as soon as possible because this PR becomes larger as time passes 😢

Comment (tsein-stripe) [reply views.go:26]: Yeah it's very much a moving target 😂

In fact my above snippet was wrong, or it only works for text input. Here's what I should have used instead:

```
type ViewState struct {
	Values map[string]map[string]BlockAction `json:"values"`
}
```

`BlockAction` already existed and handles unmarshalling for all block actions (including selects, text, date pickers, etc.)

File: views.go
Comment (tsein-stripe) [inline views.go:15]: All of the fields in this struct should be `omitempty` with the exception of `Type`, `Title`, and `Blocks` which are required View payload fields.
https://api.slack.com/reference/surfaces/views

This should allow usage of the View struct in push/update View API calls

Comment (kanata2) [reply views.go:15]: For the sake of avoiding excessive implementation sharing, I created `ModalViewRequest` and `HomeTabViewRequest` indivisually at first.
And then the role of `View` is only an entity in response from Slack.

What do you think?

Comment (tsein-stripe) [reply views.go]: Ah okay that makes sense. Well in that case, wouldn't we still wan't to omit-empty the non-required fields on the `ModalViewRequest` and `HomeTabViewRequest` structs? It might not be necessary, but I have noticed that the Slack API can be extremely picky about included fields in a payload, e.g. leaving `CallbackID` unset will actually send the Slack API `"callback_id": ""`, which could cause it to blow up, where as just omitting the field makes it happy as a clam.

Comment (tsein-stripe) [reply views.go:15]: Ahh I didn't see that, that makes perfect sense. Thank you.

**Comment by tsein-stripe** [comment]
@kanata2 I've been sitting on a set of local changes I've been making to your branch to support view interactions, I've got a PR to your fork if you'd like to include them in this PR. If not, feel free to ignore the them and I'll create a new PR against `nlopes/slack` when this PR is merged in.

https://github.com/kanata2/slack/pull/1

**Comment by russellseymour** [comment]
Hello @kanata2 @tsein-stripe 

Just wondering if this might be merged this week.

Thanks for all the work on this.

**Comment by tsein-stripe** [comment]
@kanata2 your changes LGTM. As I mentioned earlier, if you'd like we can include my changes (https://github.com/kanata2/slack/pull/1) in this PR, or I can just merge it in after your changes have been merged in.

**Review by tsein-stripe — APPROVED** [review]
LGTM w/ slight caveat -- I'm not a regular contributor to this repo

**Comment by logand22** [comment]
Also new here, It looks like the InteractionCallback Struct, could also include the new View Struct as Slack is now sending view with Interaction payload.

[Interaction Struct](https://github.com/nlopes/slack/blob/master/interactions.go#L30)

**Comment by kanata2** [comment]
@tsein-stripe Thank you for your review and [additional changes](https://github.com/kanata2/slack/pull/1). I think that you might want to submit PR to this repo after merging this!

@james-lawrence Can you review this ...? If you are busy, I want you to add someone as collaborator...😢

**Comment by tsein-stripe** [comment]
> Also new here, It looks like the InteractionCallback Struct, could also include the new View Struct as Slack is now sending view with Interaction payload.
> 
> [Interaction Struct](https://github.com/nlopes/slack/blob/master/interactions.go#L30)

That's taken care of in the PR I have against @kanata2 's branch. We just need to get this PR merged into master, then I'll submit a PR for my changes (https://github.com/kanata2/slack/pull/1) against nlopes/slack.

**Comment by ollieparsley** [comment]
Who can merge this? Would love to get this in before Xmas so I'm not building my code off a separate fork :)

**Comment by ollieparsley** [comment]
@james-lawrence are you able to look at this? Or maybe find someone who can merge?

**Comment by sturfee-petrl** [comment]
Would it be possible to add an example also?

**Comment by sfzylad** [comment]
@nlopes Any chance to get it merged?

**Comment by zpdunlap** [comment]
Any progress on this getting reviewed/merged?

**Comment by parsley42** [comment]
@kanata2 I've merged #638 and #650 to the new [slack-go/slack](https://github.com/slack-go/slack) fork that Norberto created. I tried to merge this PR as well, but got a fail in `go test -v ./...`. Would you mind having a look and submitting a PR against the new fork? I can and will merge if all tests pass (though we haven't set up Travis for it yet), and I'll go ahead and tag as v0.6.2. If things go well, that fork will replace the original nlopes/slack, I'll bump the next tag there to v0.7.0, and we'll have a larger pool of committers going forward.

Related - I'm interested in hearing any recommendations for the initial pool of committers. I have no interest in reviewing / merging all the PRs - I asked Norberto for the new org/fork explicitly so I could grow the pool of committers and make this a more scalable process. "Me" is a fine answer if you're interested, I'd just like to hear the expected "yes, definitely" from other contributors like @jojo43 or @tsein-stripe (same goes for both of you as well, naturally). Thanks.

**Comment by kanata2** [comment]
@parsley42 
Thanks for your great job!

> I tried to merge this PR as well, but got a fail in go test -v ./....
> Would you mind having a look and submitting a PR against the new fork? 

Okey, I'll check it and send PR to the new fork, but I might have take a little time.

> so I could grow the pool of committers and make this a more scalable process.

I'm interested in commiter for the repo. 
(In addition to, I want to recommend @tsein-stripe for a commiter.)

**Comment by parsley42** [comment]
Ok, not sure what my issue was before, but indeed all checks are passing, and I've re-configured Travis-CI. I'll leave this open until we hear back from @james-lawrence on the fate of this repo.

**Comment by dcelasun** [comment]
Unfortunately, this PR didn't include everything from the [kanata2/slack PR](https://github.com/kanata2/slack-old/pull/1), so the interaction stuff (e.g `InteractionTypeViewSubmission`) didn't make it :(

**Comment by kanata2** [comment]
As I mentioned in [this comment](https://github.com/slack-go/slack/pull/632#discussion_r348713375), this PR is minimal support for Views API at the time.
 If you already have implemented more, please feel free to open PR :)

**Comment by dcelasun** [comment]
Well, the PR is already in [your repository](https://github.com/kanata2/slack-old/pull/1/files) from a different author, wouldn't it be weird for *me* to send it here?

I realize that PR has more than just the rest of the Views API, so if you don't want to merge that, perhaps I could extract the view interactions from that one and make a new PR.

**Comment by kanata2** [comment]
> wouldn't it be weird for me to send it here?

Yeah, you right.
Okay, I'll check https://github.com/kanata2/slack/pull/1 and then merge if everything is fine.

**Comment by tsein-stripe** [comment]
If it's not too much effort, I don't mind you sending it here -- we've been using the changes off of my fork with a lot of success. I think extracting the view interaction changes from the original PR might be more difficult than just bringing in my changes wholesale, which includes several other block kit UI improvements, e.g. input, file, and multi select block elements.

**Comment by dcelasun** [comment]
> If it's not too much effort, I don't mind you sending it here

Thanks, but it looks like @kanata2 is about to sort it out :)

> we've been using the changes off of my fork with a lot of success

Likewise, we've been using a private fork based on your PR for a few weeks now with no issues. If that PR is merged as-is, we'd no longer need the fork.

**Comment by kanata2** [comment]
> If it's not too much effort, I don't mind you sending it here 

I think this way is better if you have the time, but of course I can also check [your PR against my fork](https://github.com/kanata2/slack/pull/1).
What do you think? @tsein-stripe

**Comment by tsein-stripe** [comment]
@kanata2 I did the git hokey pokey, pulled the upstream changes from your fork and rebased my branch on them, so it should merge cleanly now:
https://github.com/kanata2/slack/pull/1

I see there were some upstream changes around the areas I worked on, so in this PR I just overwrote them with the work from my branch. I haven't had time to test this yet.

**Comment by tsein-stripe** [comment]
Alright, @kanata2  I just finished merging in the upstream changes. I've tested the changes locally with some of our workflows and nothing looks broken, but can you and @dcelasun please give this PR a thorough review before more upstream changes get stacked on?

https://github.com/kanata2/slack/pull/1

Unfortunately this PR is still against @kanata2  fork -- GitHub won't let me create more than one fork of the slack repo. But if you're comfortable with this, we can get it merged into your fork, then submit a PR from your fork (with these changes) to slack-go.

Otherwise I'll have to delete my fork and re-fork from slack-go (which is fine with me too -- I just want to make sure I don't surprise anyone who might be relying on my fork of your fork)

**Comment by dcelasun** [comment]
Sure thing, I'll switch our fork to your updated PR tomorrow and test it internally.

**Comment by tsein-stripe** [comment]
@dcelasun How does it look? Also could you make me a contributor on this repository? We use this package pretty heavily here and it would be nice to keep the upstream repo up to date with our contributions.

**Comment by dcelasun** [comment]
I've replied on the [other PR](https://github.com/kanata2/slack/pull/1) and it looks good! I'm not an admin on this repo so I can't help with access, but @kanata2 can.

**Comment by parsley42** [comment]
@tsein-stripe Can you submit a PR against slack-go/slack, which @kanata2 (or another of us) can merge? Also, @dcelasun - do you want to recommend adding @tsein-stripe ?

**Comment by dcelasun** [comment]
@parsley42 yes, I've been using his work on view interactions and new blocks for a while now and he'd definitely be a good addition to the project.

**Comment by parsley42** [comment]
Ok - invite sent. Thanks!

**Pull request merged**
